src/js/blocks/form-fields/inspect.js

```diff
@@ -209,58 +210,99 @@ export default class Inspector extends Component {
+	 */
+	getValidationErrText( key, val ) {
+		let msg = '';
+
+		switch ( key ) {
+			case 'data_store_key':
+				// Translators: %s = user-submitted value.
+				msg = __(
+					'The user meta key "%s" is not unique. Please choose a value.',
+					'lifterlms'
+				);
+				break;
+
+			case 'id':
+				// Translators: %s = user-submitted value.
+				msg = __(
+					'The ID "%s" is not unique. Please choose a unique field ID.',
+					'lifterlms'
+				);
+				break;
+
+			case 'name':
+				// Translators: %s = user-submitted value.
+				msg = __(
+					'The name "%s" is not unique. Please choose a globally unique field name.',
+					'lifterlms'
+				);
+				break;
+
+			default:
+				// Translators: %s = user-submitted value.
+				msg = __( 'The chosen value "%s" is invalid.', 'lifterlms' );
+		}
+
+		return sprintf( msg, val );
+	}
+
+	/**
+	 * Set attributes with built-in validation
+	 *
+	 * Validates fields for uniqueness against field data in the llms/user-info-fields data
+	 * store. Prevents storage of invalid data and throws an error notice when validation issues
+	 * are encountered.
 	 *
 	 * @since [version]
 	 *
-	 * @param {string} newName The new field name
+	 * @param {string} key      Field key.
+	 * @param {string} newValue Field value.
+	 * @param {string} context  Validation context. Accepts "global" to validate against all known fields
+	 *                          or "local" to validate against loaded fields in the current form.
 	 * @return {void}
 	 */
-	updateFieldNameAttribute( newName ) {
-		const { attributes, setAttributes } = this.props,
-			currentName = attributes.name;
-
-		// We don't have to do anything here.
-		if ( newName === currentName ) {
+	updateValueWithValidation( key, newValue, context = 'local' ) {
+		const { clientId, attributes, setAttributes } = this.props,
+			currentValue = attributes[ key ],
+			{ editField, renameField } = dispatch( fieldsStore ),
+			{ createErrorNotice, removeNotice } = dispatch( 'core/notices' ),
+			noticeId = `llms-${ key }-validation-err-${ clientId }`;
+
+		// No change.
+		if ( newValue === currentValue ) {
 			return;
 		}
 
-		const { userInfoFields } = window.llms,
-			isValid = ! userInfoFields
-				.map( ( { name } ) => name )
-				.includes( newName );
+		const { getFieldBy } = select( fieldsStore ),
+			isValid = getFieldBy( newValue, key, context ) ? false : true;
 
+		removeNotice( noticeId );
 		if ( ! isValid ) {
-			const noticeId = `llms-name-validation-err-${ attributes.uuid }`,
-				{ createErrorNotice, removeNotice } = dispatch(
-					'core/notices'
-				);
-
-			removeNotice( noticeId );
-			createErrorNotice(
-				__( 'Please choose a unique field name.', 'lifterlms' ),
-				{ id: noticeId }
-			);
+			createErrorNotice( this.getValidationErrText( key, newValue ), {
+				id: noticeId,
+			} );
 
-			return;
+			// Still run updates but keep the new value valid by stripping the last character.
+			newValue = newValue.slice( 0, -1 );
```

Comment:
I know what I'm going to say it's not that important but just a note about this last char stripping.
Say I have a fields with name "foo-1", then I add another field and want to have its name like "foo-1-1" I can't do that because as soon as I write "foo-1" 1 gets stripped... so I need to do something like "foo--1" then move back with the left arrow key and add the "1" between the two dashes.
Again, this is not that important at the moment. It's just an UX we can maybe improve in the future.

Reply:
mmm we still have a problem though... I think we need to *change the name* somehow... would appending `-${n}` to the end be acceptable (in favor of stripping the last char)

Reply:
This is specifically only when changing the field's `name` attribute (used as the globally unique field ID) -- if we're trying to create a field with the name `foo-1` (and we already have a field named `foo`) and we save `foo` (temporarily) we have no way of knowing that we didn't mean to override the field without some sort of additional state management.

So I propose instead of stripping with just append `-{$n}` I'm writing the code now

Reply:
I see, alright, mmm.
Since we're now locking the post saving what's the problem in not changing the name?
You say:
> mmm we still have a problem though...

Which one?